packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1502,6 +1502,68 @@ abstract class ResidentRunner extends ResidentHandlers {
     appFinished();
   }
 
+  @protected
+  Future<void> performPreHotRestart() async {
+    final Timer slowCallbackTimer = Timer(const Duration(seconds: 5), () async {
+      globals.printError('preHotRestartCallbacks are taking longer than expected...');
+    });
+    final List<Future<void>> preHotRestartFutures = <Future<void>>[];
+    for (final FlutterDevice device in flutterDevices) {
+      final FlutterVmService? vmService = device.vmService;
+      if (vmService == null) {
+        continue;
+      }
+      final List<FlutterView> views = await vmService.getFlutterViews();
+      for (final FlutterView view in views) {
+        final String? isolateId = view.uiIsolate?.id;
+        if (isolateId == null) {
+          continue;
+        }
+
+        // Make sure that we are able to run the preHotRestart callbacks,
+        // by unpausing the isolate and removing breakpoints, otherwise we will be
+        // infinitely waiting for the callbacks to run.
+        await unpauseIsolateAndRemoveBreakpoints(vmService, isolateId);
+
+        preHotRestartFutures.add(vmService.flutterInvokePreHotRestartCallbacks(
+          isolateId: isolateId,
+        ));
+      }
+    }
+    await Future.wait(preHotRestartFutures);
```

Comment:
my biggest concern with this PR is this line; it introduces a potential point where hot restart will fail and there's no way around it (short of ^C and rerunning the app).

Reply:
If an exception is captured while running the callback, the user will know from the command line and the app will restart normally.  Or are you referring to another type of error like an infinite loop?
Here is an example of a callback that fails:
![image](https://github.com/flutter/flutter/assets/22084723/ec460b5b-31cd-491d-aa82-17cbfb1bd82e)

Reply:
what if the future just doesn't complete?